File: PROG/Week02/Lab03/NumberToWords.java
import java.util.HashMap;
import java.util.Map;

public class NumberToWords {
    static Map<Integer, String> ones_word = new HashMap<Integer, String>() {
        {
            put(1, "one");
            put(2, "two");
            put(3, "three");
            put(4, "four");
            put(5, "five");
            put(6, "six");
            put(7, "seven");
            put(8, "eigth");
            put(9, "nine");
        }
    };
    static Map<Integer, String> tens_word = new HashMap<Integer, String>() {
        {
            put(1, "ten");
            put(2, "twenty");
            put(3, "thirty");
            put(4, "forty");
            put(5, "fifty");
            put(6, "sixty");
            put(7, "seventy");
            put(8, "eighty");
            put(9, "ninety");
        }
    };
    static Map<Integer, String> teens_word = new HashMap<Integer, String>() {
        {
            put(10, "ten");
            put(11, "eleven");
            put(12, "twelve");
            put(13, "thirteen");
            put(14, "fourteen");
            put(15, "fifteen");
            put(16, "sixteen");
            put(17, "seventeen");
            put(18, "eighteen");
            put(19, "nineteen");
        }
    };

    public static String intToWords(int num) {
        String word = "";
        int hundreths = num / 100;
        int tens = (num / 10) % 10;
        int ones = num % 10;
        int teen = num % 100;

        if (hundreths != 0) {
            word += ones_word.get(hundreths) + " hundred ";
            if (tens != 0 || ones != 0)
                word += " and ";
        }

        if (tens > 1)
            word += " " + tens_word.get(tens) + " ";

        if (tens == 1)
            word += " " + teens_word.get(teen) + " ";

        if (ones != 0 && tens != 1)
            word += " " + ones_word.get(ones) + " ";

        if (word == "")
            word = "zero";

        return word.replaceAll(" +", " ").trim();
    }

    public static String intToRoman(int num) {
        String word = "";

        while (num > 0) {
            if (num >= 900) {
                word += "CM";
                num -= 900;
            } else if (num >= 500) {
                word += "D";
                num -= 500;
            } else if (num >= 400) {
                word += "CD";
                num -= 400;
            } else if (num >= 100) {
                word += "C";
                num -= 100;
            } else if (num >= 90) {
                word += "XC";
                num -= 90;
            } else if (num >= 50) {
                word += "L";
                num -= 50;
            } else if (num >= 40) {
                word += "XL";
                num -= 40;
            } else if (num >= 10) {
                word += "X";
                num -= 10;
            } else if (num >= 9) {
                word += "IX";
                num -= 9;
            } else if (num >= 5) {
                word += "V";
                num -= 5;
            } else if (num >= 4) {
                word += "IV";
                num -= 4;
            } else {
                word += "I";
                num -= 1;
            }
        }

        return word;
    }

    public static int readInt() {
        System.out.print("Number: ");
        return In.nextInt();
    }

    public static void main(String[] args) {
        int n;
        while ((n = readInt()) != -1) {
            String words = intToWords(n);
            String roman = intToRoman(n);
            System.out.println("In Words: " + words);
            System.out.println("In Roman: " + roman);
        }

        System.out.println("Done");
    }
}

File: PROG/Week03/Lab03/NumberToWords.py
import re

ones_word = {
    1: "one",
    2: "two",
    3: "three",
    4: "four",
    5: "five",
    6: "six",
    7: "seven",
    8: "eigth",
    9: "nine",
}
tens_word = {
    1: "ten",
    2: "twenty",
    3: "thirty",
    4: "forty",
    5: "fifty",
    6: "sixty",
    7: "seventy",
    8: "eighty",
    9: "ninety",
}
teens_word = {
    10: "ten",
    11: "eleven",
    12: "twelve",
    13: "thirteen",
    14: "fourteen",
    15: "fifteen",
    16: "sixteen",
    17: "seventeen",
    18: "eighteen",
    19: "nineteen",
}


def int_to_words(n):
    word = ""
    hundreths = n // 100
    tens = (n // 10) % 10
    ones = n % 10
    teen = n % 100

    if hundreths != 0:
        word += ones_word[hundreths] + " hundred "
        if tens != 0 or ones != 0:
            word += " and "

    if tens > 1:
        word += f" {tens_word[tens]} "

    if tens == 1:
        word += f" {teens_word[teen]} "

    if ones != 0 and tens != 1:
        word += f" {ones_word[ones]} "

    if word == "":
        word = "zero"

    return re.sub(" +", " ", word).strip()


def int_to_roman(num):
    """it's fucking beatiful"""
    word = ""

    while num > 0:
        if num >= 900:
            word += "CM"
            num -= 900
        elif num >= 500:
            word += "D"
            num -= 500
        elif num >= 400:
            word += "CD"
            num -= 400
        elif num >= 100:
            word += "C"
            num -= 100
        elif num >= 90:
            word += "XC"
            num -= 90
        elif num >= 50:
            word += "L"
            num -= 50
        elif num >= 40:
            word += "XL"
            num -= 40
        elif num >= 10:
            word += "X"
            num -= 10
        elif num >= 9:
            word += "IX"
            num -= 9
        elif num >= 5:
            word += "V"
            num -= 5
        elif num >= 4:
            word += "IV"
            num -= 4
        else:
            word += "I"
            num -= 1

    return word


if __name__ == "__main__":
    while (n := int(input("Number: "))) != -1:
        words = int_to_words(n)
        roman = int_to_roman(n)
        print(f"In Words: {words}")
        print(f"In Roman: {roman}")

    print("Done")
